WebGL2 Instanced Rendering with `drawArraysInstanced()`

adriandarian opened the pull request:
This PR implements WebGL2 instanced rendering for the WebGL drawer, allowing multiple tiles to be drawn with a single `drawArraysInstanced()` call instead of the current batched approach that draws 6 vertices per tile.

### Changes

#### New Instanced Rendering Path
- Added `_makeFirstPassInstancedShaderProgram()` to create GLSL ES 3.0 shaders for instanced rendering
- Added `_drawTilesInstanced()` method that uses `gl.drawArraysInstanced()` to render tiles
- Added `_getTileDataInstanced()` to prepare per-instance attribute data
- Refactored original drawing code into `_drawTilesBatched()` for clarity and fallback support

#### GLSL ES 3.0 Shaders
- Vertex shader uses `in`/`out` syntax with per-instance attributes
- Fragment shader generates texture lookup conditionals at compile time to work around GLSL sampler array indexing limitations
- Transform matrix passed as 3 separate `vec3` columns to avoid mat3 vertex attribute compatibility issues

#### Per-Instance Attributes
Using `vertexAttribDivisor()` to advance once per instance:
- `a_texture_position_tl` (vec2) - Texture top-left coordinates
- `a_texture_position_size` (vec2) - Texture dimensions
- `a_transform_col0/1/2` (vec3 × 3) - Transform matrix columns
- `a_opacity` (float) - Tile opacity
- `a_texture_index` (float) - Texture unit index

#### Graceful Fallback
- Automatically falls back to batched rendering on WebGL1 contexts
- Try-catch wrapper ensures fallback if shader compilation fails
- Properly resets vertex attribute divisors after drawing to prevent state conflicts with other shaders

### Technical Notes

**Vertex Attribute Divisors are Global State**
A key discovery during implementation: `vertexAttribDivisor()` sets state per attribute location globally, not per shader program. This required:
1. Setting divisors to 1 before each instanced draw call
2. Resetting divisors to 0 after drawing to prevent conflicts with the second pass shader and batched first pass shader

**Sampler Array Indexing Limitation**
GLSL ES 3.0 requires constant integral expressions for sampler array indices. The fragment shader generates explicit if/else conditionals at compile time rather than using dynamic loop indexing.

### Testing
All 311 existing tests pass.

### Related
- WebGL2 context support (prerequisite)
- Future optimization: VAO (Vertex Array Objects) could further reduce state changes

iangilman commented:
@adriandarian I haven't heard from you in a while! Are you still interested in continuing to work on these patches? 

I'm going to mark this as "up for grabs"... anyone who wants to can pick it up and take it over the finish line. That includes @adriandarian of course if you want to come back to it 😊

adriandarian committed:
feat(webgl2): implement instanced rendering with drawArraysInstanced()

- Add WebGL2 instanced rendering path for improved performance
- Create GLSL ES 3.0 shaders for instanced rendering
- Use vertexAttribDivisor for per-instance attributes:
  - Texture position (TL and size)
  - Transform matrix (as 3 vec3 columns)
  - Opacity and texture index
- Generate texture lookup conditionals at compile time to work
  around GLSL ES 3.0 sampler array indexing limitations
- Properly manage vertex attribute divisors (global state)
  to avoid conflicts with other shaders
- Fall back to batched rendering if instanced setup fails
- Separate _drawTilesInstanced and _drawTilesBatched methods

All 311 tests pass.

adriandarian committed:
Fix WebGL tile batching and instanced draw cleanup

iangilman reviewed (commented):
@adriandarian Thank you for updating this PR and getting back on it!

This certainly seems like a good idea, and I would hope it is a nice improvement to performance. Any thoughts about how we might quantify that? I feel we don't really have good performance tests.

Any thoughts about the Copilot comments? Are any of them meaningful? Would you say you've addressed them?

@pearcetm @Aiosa Any thoughts about this PR? The general structure looks reasonable to me, but I'm a bit out of my depth with the specifics of WebGL.

adriandarian committed:
Merge branch 'master' into feat/webgl2-instanced-rendering

adriandarian committed:
test(webgl): align drawer internal tests with gl context

adriandarian commented on src/webgldrawer.js:
```diff
@@ -606,9 +601,232 @@
+                    // We've filled up the buffers: time to draw this set of tiles
+
+                    // bind each tile's texture to the appropriate gl.TEXTURE#
+                    for(let i = 0; i <= numTilesToDraw; i++){
```
Addressed on the current branch. _bindTileTextures now iterates with i < drawCount, so it only binds valid texture units and does not step past maxTextures.

adriandarian commented on src/webgldrawer.js:
```diff
@@ -78,7 +78,32 @@
+             * - Transform feedback: GPU-based computations for tile processing
+             * - Multiple render targets: Advanced compositing with multiple framebuffer attachments
+             * - GLSL ES 3.0 shaders: Enhanced shader capabilities (integer ops, etc.)
+             *
```
Addressed. The TODO block has been updated so instanced rendering is no longer listed as pending and is now called out as an implemented WebGL2 feature.

adriandarian commented on test/modules/drawer.js:
```diff
@@ -459,4 +459,225 @@
+                aTexturePosition: 'aTexture',
+                aIndex: 'aIndex',
+                uTransformMatrices: ['m0', 'm1'],
+                uOpacities: 'uOpacities'
+            },
+            getDataToDraw(tile) {
+                return tile.skip ? null : {texture: tile.texture};
+            },
+            _getTileData(tile, tiledImage, textureInfo, overallMatrix, index, texturePositionArray, textureDataArray, matrixArray, opacityArray) {
+                texturePositionArray.set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], index * 12);
+                textureDataArray[index] = textureInfo.texture;
+                matrixArray[index] = [index + 1, 0, 0, 0, index + 1, 0, 0, 0, 1];
+                opacityArray[index] = 1;
+            },
+            _bindTileTextures: OpenSeadragon.WebGLDrawer.prototype._bindTileTextures
+        };
```
Addressed in a373ab5d. This test now provides a fake _glContext with getContext and getFirstPass, so _drawTilesBatched exercises the refactored code path instead of returning early.

adriandarian commented on src/webgldrawer.js:
```diff
@@ -606,9 +601,232 @@
+                }
+
+                if( (numTilesToDraw === maxTextures) || (tileIndex === tilesToDraw.length - 1)){
+                    // We've filled up the buffers: time to draw this batch of tiles
```
Addressed in the current instanced path. The code now uses drawCount as the instance index and only increments it for tiles with valid textureInfo.texture, so the uploaded instance buffers and bound textures stay aligned with the actual draw count.

adriandarian commented on src/webgldrawer.js:
```diff
@@ -606,9 +601,232 @@
+
+            // iterate over tiles and add data for each one to the buffers
+            for(let tileIndex = 0; tileIndex < tilesToDraw.length; tileIndex++){
+                const tile = tilesToDraw[tileIndex].tile;
+                const indexInDrawArray = tileIndex % maxTextures;
+                const numTilesToDraw =  indexInDrawArray + 1;
+                const textureInfo = this.getDataToDraw(tile);
+
+                if (textureInfo && textureInfo.texture) {
+                    this._getTileData(tile, tiledImage, textureInfo, overallMatrix, indexInDrawArray, texturePositionArray, textureDataArray, matrixArray, opacityArray);
+                }
```
Addressed in the current implementation. The batched path now uses drawCount as the write index and draw count, and only increments it when getDataToDraw returns a valid texture, so skipped tiles do not contribute uninitialized matrix, opacity, or texture state.

adriandarian commented on test/modules/drawer.js:
```diff
@@ -459,4 +459,225 @@
+            aTextureIndex: 'aTextureIndex',
+            aOutputPosition: 'aOutput'
+        };
+        const drawer = {
+            _gl: gl,
+            _firstPassInstanced: pass,
+            _glNumTextures: 3,
+            _textureUnitIndices: null,
+            _instancedArrays: null,
+            _instancedArrayCapacity: 0,
+            getDataToDraw(tile) {
+                return tile.skip ? null : {texture: tile.texture};
+            },
+            _getTextureUnitIndices: OpenSeadragon.WebGLDrawer.prototype._getTextureUnitIndices,
+            _ensureInstancedArrays: OpenSeadragon.WebGLDrawer.prototype._ensureInstancedArrays,
+            _bindTileTextures: OpenSeadragon.WebGLDrawer.prototype._bindTileTextures,
+            _resetInstancedAttributeDivisors: OpenSeadragon.WebGLDrawer.prototype._resetInstancedAttributeDivisors,
+            _getTileDataInstanced(tile, tiledImage, textureInfo, overallMatrix, index, texturePositionTLArray, texturePositionSizeArray, transformMatrixArray, opacityArray, textureIndexArray, textureDataArray) {
+                texturePositionTLArray.set([0, 0], index * 2);
+                texturePositionSizeArray.set([1, 1], index * 2);
+                transformMatrixArray.set([1, 0, 0, 0, 1, 0, 0, 0, 1], index * 9);
+                opacityArray[index] = 1;
+                textureIndexArray[index] = index;
+                textureDataArray[index] = textureInfo.texture;
+            }
+        };
```
Addressed in a373ab5d. The instanced test now stubs _glContext.getContext so _drawTilesInstanced and its helpers run through the actual WebGL2 path, including the divisor reset behavior that the test asserts.

adriandarian commented on test/modules/drawer.js:
```diff
@@ -459,4 +459,225 @@
+            vertexAttribDivisor() {},
+            getAttribLocation() { return 0; },
+            getUniformLocation() { return {}; },
+            useProgram() {},
+            uniform1iv() {}
+        };
+        const drawer = {
+            _isWebGL2: true,
+            _gl: gl,
+            _glNumTextures: 2,
+            _unitQuad: new Float32Array([0, 0, 1, 0, 0, 1]),
+            _textureUnitIndices: null,
+            _firstPassInstanced: {stale: true},
+            _getTextureUnitIndices: OpenSeadragon.WebGLDrawer.prototype._getTextureUnitIndices,
+            _resetInstancedAttributeDivisors: OpenSeadragon.WebGLDrawer.prototype._resetInstancedAttributeDivisors
+        };
```
Addressed in a373ab5d. The shader setup test now supplies a minimal _glContext with getContext, getMaxTextures, and getUnitQuad, so the fallback is triggered by the mocked shader compilation failure rather than a missing-context TypeError.

adriandarian commented on src/webgldrawer.js:
```diff
@@ -606,9 +601,232 @@
+            const transformMatrixArray = new Float32Array(maxTextures * 9);     // mat3 per instance
+            const opacityArray = new Float32Array(maxTextures);                 // float per instance
+            const textureIndexArray = new Float32Array(maxTextures);            // float per instance
+            const textureDataArray = new Array(maxTextures);
```
Addressed. The instanced path now reuses cached typed arrays via _ensureInstancedArrays and reuses the sampler index array via _getTextureUnitIndices, so those allocations are not rebuilt on every draw call.

adriandarian commented on src/webgldrawer.js:
```diff
@@ -606,9 +601,232 @@
+                    Math.min(4, this._maxAnisotropy)
+                );
+            }
+
```
Addressed. getTextureFilter is now pure, and anisotropy is applied explicitly in the texture setup sites where the intended texture is already bound.

adriandarian committed:
Merge branch 'master' into feat/webgl2-instanced-rendering

adriandarian committed:
Merge branch 'master' into feat/webgl2-instanced-rendering

adriandarian committed:
fix(ci): raise raw bundle size limit

adriandarian committed:
fix(ci): clear puppeteer cache for base bundle build
